Initial Classes

mecha opened the pull request:
The initial classes. Or more specifically, traits.

This PR introduces **four** main traits for PDO based resource models, as well as various other traits that each provide a fragment of the implementation of the query building process.

1. [`PdoSelectCapableTrait`]
2. [`PdoInsertCapableTrait`]
3. [`PdoUpdateCapableTrait`]
4. [`PdoDeleteCapableTrait`]

[`AbstractPdoResourceModel`] is also provided for basic PDO awareness and query execution functionality.

[`PdoSelectCapableTrait`]: https://github.com/RebelCode/cqrs-resource-model-abstract/blob/task/initial-classes/src/Pdo/PdoSelectCapableTrait.php
[`PdoInsertCapableTrait`]: https://github.com/RebelCode/cqrs-resource-model-abstract/blob/task/initial-classes/src/Pdo/PdoInsertCapableTrait.php
[`PdoUpdateCapableTrait`]: https://github.com/RebelCode/cqrs-resource-model-abstract/blob/task/initial-classes/src/Pdo/PdoUpdateCapableTrait.php
[`PdoDeleteCapableTrait`]: https://github.com/RebelCode/cqrs-resource-model-abstract/blob/task/initial-classes/src/Pdo/PdoDeleteCapableTrait.php

[`AbstractPdoResourceModel`]: https://github.com/RebelCode/cqrs-resource-model-abstract/blob/task/initial-classes/src/Pdo/AbstractPdoResourceModel.php

mecha committed:
Installed required dependencies

mecha committed:
Added `BuildSqlWhereCapableTrait` and test

mecha committed:
Added `EscapeSqlReferenceCapableTrait` and test

mecha committed:
Added `BuildSqlJoinsCapableTrait` and test

mecha committed:
Added `BuildDeleteSqlCapableTrait` and test

mecha committed:
Added `BuildInsertSqlCapableTrait` and test

mecha committed:
Added `BuildSelectSqlQueryCapableTrait` and test

mecha committed:
Added `ExecutePdoQueryCapableTrait` and test

mecha committed:
Added `GetPdoExpressionHashMapCapableTrait` ...

... and test.

mecha committed:
Added `PdoAwareTrait` and test

mecha committed:
Documented exceptions for `_renderSqlCondition()`

mecha committed:
Added `RenderSqlQueryCapableTrait` and test

mecha committed:
Added `AbstractPdoResourceModel`

mecha committed:
Improved INSERT SQL building ...

... The `_buildInsertSql()` method now accepts the column map and
value hashmap, which are used to map fields to columns and turn
values into hashes.

XedinUnknown commented on src/Pdo/PdoAwareTrait.php:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    /**
+     * Creates a new Dhii invalid argument exception.
```
Remove the "Dhii".

XedinUnknown commented on src/Pdo/Query/BuildInsertSqlCapableTrait.php:
```diff
@@ -0,0 +1,179 @@
+trait BuildInsertSqlCapableTrait
+{
+    /**
+     * Builds a INSERT SQL query.
```
"a" -> "an".

XedinUnknown commented on src/Pdo/Query/BuildInsertSqlCapableTrait.php:
```diff
@@ -0,0 +1,179 @@
+use InvalidArgumentException;
+
+/**
+ * Common functionality for objects that can built INSERT SQL queries.
```
"built" -> "build"

XedinUnknown commented on src/Pdo/Query/BuildDeleteSqlCapableTrait.php:
```diff
@@ -0,0 +1,70 @@
+    protected function _buildDeleteSql(
+        $table,
+        LogicalExpressionInterface $condition = null,
+        array $columnMap = [],
```
Why is it the concern of these methods to deal with field-to-column mapping?

XedinUnknown commented on src/Pdo/Query/BuildInsertSqlCapableTrait.php:
```diff
@@ -0,0 +1,179 @@
+        $table,
+        array $columns,
+        array $rowSet,
+        array $columnMap = [],
```
Why is it the concern of these methods to deal with field-to-column mapping?

XedinUnknown commented on src/Pdo/Query/BuildInsertSqlCapableTrait.php:
```diff
@@ -0,0 +1,179 @@
+    protected function _buildSqlValuesList(
+        array $columns,
+        array $rowSet,
+        array $columnMap = [],
```
Why is it the concern of these methods to deal with field-to-column mapping?

XedinUnknown commented on src/Pdo/Query/BuildInsertSqlCapableTrait.php:
```diff
@@ -0,0 +1,179 @@
+            $values[] = $this->_buildSqlRowValues($columns, $_rowData, $columnMap, $valueHashMap);
+        }
+
+        return sprintf('VALUES %s', implode(', ', $values));
```
What do you think about leaving the `VALUES` word itself up to the consumer of this method?

XedinUnknown commented on src/Pdo/Query/BuildInsertSqlCapableTrait.php:
```diff
@@ -0,0 +1,179 @@
+    protected function _buildSqlRowValues(
+        array $columns,
+        array $rowData,
+        array $columnMap = [],
```
Why is it the concern of these methods to deal with field-to-column mapping?

mecha committed:
Fixed typos in docs

XedinUnknown commented:
1. Column map is an implementation detail. Furthermore, it's a configuration-time detail, as opposed to call-time. This is because the CQRS model knows the field-to-table mapping.
2. The expression field-to-field mapping is a different story. This can be done when building the expression, or when un-serializing a serialized expression. Eventual expressions must have the fields be final. They can later be translated by the CQRS model into a column. Therefore, the column map should no longer be accepted as method arg, but instead the method can retrieve it via a different abstract method.
3. To facilitate this translation, the column map needs to be augmented to be able to use an object as value. This object provides the column name and table name both. For example, `$map = ['age' => new EntityField('user', 'age')]`.

Eventually, the hash mapping remains abstract but becomes more concrete specifically for the PDO trait model. At the same time, the field-to-column mapping remains abstract for *any* SQL trait model, because they may chose to map or not map.

mecha committed:
Installed `dhii/sql-interface`

mecha committed:
Fixed tests typos and PHP 5.4 compatibility

mecha committed:
Updated dependencies

mecha committed:
Removed use of `$columnMap` parameter

mecha committed:
Added abstract column map retrieval method ...

... in `RenderSqlConditionCapableTrait`.

mecha committed:
Fixed code standards

mecha commented:
@XedinUnknown Take a look now. I've removed all uses of `$columnMap` parameters. The `RenderSqlConditionCapableTrait` now depends on an abstract `_getFieldColumnMap()` method. Furthermore, it expects it to return a map of [`EntityFieldInterface`](https://github.com/Dhii/sql-interface/blob/task/initial-interfaces/src/EntityFieldInterface.php) instances.

mecha committed:
Improved param docs

mecha committed:
Fixed incorrect test name

mecha committed:
Added `BuildUpdateSqlCapableTrait` and test

mecha committed:
Refactored method name to be SQL specific

mecha committed:
Added `PdoSelectCapableTrait` and test

mecha committed:
Made value hashing method abstract

mecha committed:
PHP doc grammar fixes

mecha committed:
Added comments for added clarity

mecha committed:
Added `GetPdoValueHashStringCapableTrait` and test

mecha committed:
Empty row set now throws exception

mecha committed:
Added `PdoInsertSqlCapableTrait` and test

mecha committed:
Added array type hint

mecha committed:
Improved specificity of PHP method doc

mecha committed:
Fixed mock callback method mocking

mecha committed:
Added missing `throws` doc annotation

mecha committed:
Disabled SQL inspections in PHPStorm

mecha committed:
Added `PdoInsertCapableTrait#insert()` assertion

mecha committed:
Fixed double PDO INSERT query execution

mecha committed:
Fixed test for `PdoInsertCapableTrait#insert()`

mecha committed:
Added missing `return` doc annotation

mecha committed:
Fixed optional params in ...

... `BuildUpdateSqlCapableTrait#_buildUpdateSql()`

mecha committed:
Added `PdoUpdateCapableTrait` and test

mecha committed:
Fixed uses of `Stringable`

mecha committed:
Added `PdoDeleteCapableTrait` and test

mecha committed:
Removed use of PDO SqLite in tests

- Since the DB and PDO are not the SUT, they have now been replaced with mocked instances
- Given that most of the query-building functionality is mocked anyway, testing real database actions meant mostly testing the mocks.

mecha committed:
Removed SQL reference escaping functionality ...

... from `AbstractPdoResourceModel`

mecha committed:
Fixed mocks for `EntityFieldInterface`

mecha committed:
Updated dependencies

mecha committed:
Fixed test for `BuildUpdateSql#_buildUpdateSql()`

mecha committed:
Updated dependencies

mecha committed:
Updated PDO hash map algorithm using new ...

... expressions API.

XedinUnknown commented on src/Pdo/PdoDeleteCapableTrait.php:
```diff
@@ -0,0 +1,102 @@
+     *
+     * @return string[]|Stringable[] A list of field names.
+     */
+    abstract protected function _getSqlDeleteFieldNames();
```
This method returns a map, but the name doesn't hint that.

mecha replied:
It doesn't return a map. It returns a list of field names.

XedinUnknown commented on src/Pdo/PdoDeleteCapableTrait.php:
```diff
@@ -0,0 +1,102 @@
+     */
+    protected function _delete(LogicalExpressionInterface $condition = null)
+    {
+        $fieldNames = $this->_getSqlDeleteFieldNames();
```
This variable holds a map, but the name doesn't hint that.

mecha replied:
It doesn't hold a map. It holds a list of field names.

XedinUnknown replied:
In this case, we specifically need a list of *unique* field names. Although a map probably exists somewhere else, it is not usable here.

XedinUnknown commented on src/Pdo/PdoInsertCapableTrait.php:
```diff
@@ -0,0 +1,135 @@
+     *
+     * @param array $record An associative array mapping field names to their values for a single record.
+     *
+     * @return array The generated value hash map, mapping values to their respective hashes.
```
I think the return type could be more narrow and descriptive.

XedinUnknown replied:
How about I create a new `MapInterface extends ContainerInterface, Traversable`, and you can return that?

mecha replied:
I don't think this is an adequate place to be dealing with high-level data structures. This is not API-level, but implementation logic code, that will feed its result into another non-API method.

Thoughts?

FYI this method still needs to return an array as it's required to be used in PHP array functions later.

XedinUnknown commented on src/Pdo/PdoInsertCapableTrait.php:
```diff
@@ -0,0 +1,135 @@
+     *
+     * @since [*next-version*]
+     *
+     * @param array $records A list of containers, each containing the record data to insert.
```
Why not make this a `Traversable`?

mecha replied:
Don't know why not, it's a good idea. I'll make the changes.

mecha replied:
After giving it some more thought, it makes sense to be able to accept a `Traversable` of `ContainerInterface`s here, as you suggested [here](https://github.com/RebelCode/cqrs-resource-model-abstract/pull/1#discussion_r158841594). Not only does this make sense given how we view container, but it would also allow us to pass data models directly without having to transform them into arrays first.

Alternatively, we can also accept a `Traversable` of `Traversable`s.

mecha replied:
I would say we pick something and stick to it, at least for each record, i.e. an array of traversable of _either_ sub-arrays or containers.

At this point, the only thing that's making me hesitant from using containers is the fact that consumers are then bound to using them which might mean they would have to construct them. Is this a valid concern?

mecha replied:
@XedinUnknown seems to have identified the same problem while working on something else, and has come up with a solution which I quite like. It involves depending on abstract methods which functional similar to a container's `get()` and `has()` methods, except that they have a container/array as parameter.

This means that our classes can use these methods to retrieve something by key while not knowing what they are retrieving from, be it a container, array or something else. But most importantly, it allows us to throw container-related exceptions for non-container collections, such as arrays. With traits that implement this functionality, this seems to be the best approach.

So I will make the resource model abstracts accept both arrays and container, and depend on these abstract methods for internal reading.

XedinUnknown commented on src/Pdo/PdoInsertCapableTrait.php:
```diff
@@ -0,0 +1,135 @@
+     *
+     * @since [*next-version*]
+     *
+     * @param array $record An associative array mapping field names to their values for a single record.
```
How about a `Traversable`?

mecha replied:
Don't know why not, it's a good idea. I'll make the changes.

XedinUnknown commented on src/Pdo/PdoInsertCapableTrait.php:
```diff
@@ -0,0 +1,135 @@
+     *
+     * @param string|Stringable     $table        The name of the table to insert into.
+     * @param string[]|Stringable[] $columns      A list of columns names. The order is preserved in the built query.
+     * @param array                 $rowSet       The record data as a map of column names to values.
```
`Traversable` of containers?

mecha replied:
Don't think it's possible. The input set of record data is a mapping of **field names** to values, but that needs to be transformed into a map of **column names** names to values. Unless of course you mean the input can be a `Traversable` of containers but the output remains an array.

Of course, there's also the argument of "adapting from the outside vs adapting internally".

mecha replied:
As I write this I realize that no field-to-column transformation is being done for the `$records`, but there should be.

XedinUnknown commented on src/Pdo/PdoSelectCapableTrait.php:
```diff
@@ -0,0 +1,135 @@
+     *
+     * @since [*next-version*]
+     *
+     * @param string[]|Stringable[]           $columns        A list of names of columns to select.
```
In such cases, should we perhaps also allow `Traversable`?

mecha replied:
Yes, good idea. `Traversable` fits with the way this parameter is being used.

XedinUnknown commented on src/Pdo/PdoUpdateCapableTrait.php:
```diff
@@ -0,0 +1,194 @@
+     *
+     * @since [*next-version*]
+     *
+     * @param array                           $changeSet A map of field names mapping to the values to change.
```
Can these be SQL expressions? What happens if they are? What is this map expected to contain?

mecha replied:
> Can these be SQL expressions?

I guess they can be. I considered this for a while, but I hadn't found whether expressions are supported in UPDATE queries or whether it was right to accept them given that only a few SQL databases/dialects might support them.

> What happens if they are?

This file would have more characters.

> What is this map expected to contain?

Field names mapping to new values. These are used to build the `SET` portion of the query.

```
UPDATE my_table SET fieldName = "value", otherField = "other value";
```

XedinUnknown reviewed (changes requested):
I totally love this implementation. It looks like a lot of work, thought, and normalization went in. Great job!

In terms of improvement, I think it could be a good idea to at least optionally allow containers or iterables - depending on how a method uses the parameter. I will make a new interface for maps that is an iterable container, and methods can create those.

A few minor naming changes could also be a good idea. Maybe even inline comments: I found it hard to follow some of the primary methods for CRUD.

In terms of documentation, I would like to see list/map parameter docs explain what the list/map is supposed to contain.

XedinUnknown commented:
Please add branch alias for `develop`?

mecha committed:
Aliased `develop` to `0.1.x-dev`

mecha committed:
Fixed `PdoInsssertCapableTrait` ...

- Field-to-column transformation was not implemented
- Record pre-processing now supports containers

mecha committed:
Added test for `WpSelectCapableTrait`

mecha committed:
Added `WpSelectCapableTrait`

XedinUnknown committed:
Switched `expression-interface` package to more stable version

This also fixes build, as previous build was based on an obsolete
branch.
